inference/__init__.py

```diff
@@ -0,0 +1,492 @@
+        elif sparse_compression_strategy == 1:
+            row_offsets_output = self.row_offsets.diff().reshape((-1, 16)).max(axis=1).values * 16
+            row_offsets_output = row_offsets_output.cumsum(dim=0)
+            row_offsets_output = torch.cat((torch.tensor([0]), row_offsets_output)).to(dtype=torch.int32)
+            col_val_count = row_offsets_output[-1]
+            col_vals_output = torch.zeros(col_val_count, dtype=torch.int32)
+        else:
+            raise 'Invalid sparse compression strategy requested.'
+
+        spqr_cuda.tensor_compress_interleaved(m, n, bits, W, beta1, beta2, W_s, W_z, W_s_s, W_s_z, W_z_s, W_z_z,
+                                              self.row_offsets,
+                                              row_offsets_output,
+                                              self.col_vals,
+                                              col_vals_output,
+                                              self.buff0,
+                                              sparse_compression_strategy)
+        self.row_offsets = row_offsets_output
+        self.col_vals = col_vals_output
+
+
+class SPQRModule(torch.nn.Module):
+    def __init__(self, spqr_host: SPQRUncompressed):
+        super().__init__()
+        self.m = spqr_host.m
+        self.n = spqr_host.n
+        self.bits = spqr_host.bits
+        self.beta1 = spqr_host.beta1
+        self.beta2 = spqr_host.beta2
+        self.buff0 = spqr_host.buff0
+        self.row_offsets = spqr_host.row_offsets
+        self.col_vals = spqr_host.col_vals
+        self.deq_w = None
+        self.in_perm = spqr_host.in_perm
+        self.out_perm = spqr_host.out_perm
+
+
+    def to_device(self, device: torch.device):
+        self.buff0 = self.buff0.to(device=device)
+        self.row_offsets = self.row_offsets.to(device=device)
+        self.col_vals = self.col_vals.to(device=device)
+
+    @property
+    def nnz(self):
+        return self.col_vals.shape[0]
+
+    def _apply(self, fn):
+        # Apply the function to all parameters and buffers
+        super(SPQRModule, self)._apply(fn)
+
+        # Apply the function to custom attributes
+        self.buff0 = fn(self.buff0)
+        self.row_offsets = fn(self.row_offsets)
+        self.col_vals = fn(self.col_vals)
+
+        if self.in_perm is not None:
+            self.in_perm = fn(self.in_perm)
+
+        if self.out_perm is not None:
+            self.out_perm = fn(self.out_perm)
+
+
+        return self
+
+    @property
+    def density(self) -> float:
+        return self.col_vals.shape[0] / (self.m * self.n)
+
+    @property
+    def sparsity(self) -> float:
+        return 1 - self.density
+
+    def forward(self, x: T) -> T:
+        inner_dim = x.shape[1]
+
+        y = torch.empty((inner_dim * self.m), dtype=torch.float16, device=self.buff0.device)
+
+        for i in range(inner_dim):
+            _x = x[..., i, :].flatten()
+            if self.in_perm is not None:
+                _x = _x[self.in_perm]
```

Comment:
if you are doing in_perm , don't you need to do out_perm?

Reply:
As per our meeting, the thought was to have the model permuting the output tensor instead of the input tensor. I've since decided to remove this after adding support for cuda graphs.